lib/Optimizer/GraphOptimizer/ConstantFolding.cpp

```diff
@@ -0,0 +1,277 @@
+  run(backend, *compiledF, bindings);
+  return llvm::Error::success();
+}
+
+/// Perform constant folding in the function \p F . Any non-trivial node (i.e.
+/// not a constant or a splat) that can be computed at compile-time is going to
+/// be computed at compile-time. \returns true if any foldings were performed.
+bool glow::ConstantFold::run(Function *F) {
+  LOG_SCOPE(F->getLogContext(), "glow::constantFold")
+  CompilationContext cctx;
+  // Graph optimizations may be required to e.g. perform lowering of some nodes
+  // that are not supported natively.
+  cctx.optimizationOpts.enableGraphOptz = true;
+  // Do not recursively call constant folding.
+  cctx.optimizationOpts.enableConstantFolding = false;
+  cctx.backendOpts.collectConstants = true;
+  bool changed = false;
+  // Backend to be used for compile-time computations.
+  std::unique_ptr<Backend> backend(createBackend("Interpreter"));
+  // Set of constant operations to be computed.
+  llvm::SmallVector<Node *, 64> worklist;
+  // Traverse nodes in post-order, so that children are seen before parents.
+  GraphPostOrderVisitor postOrderVisitor(*F);
+  auto nodes = postOrderVisitor.getPostOrder();
+  // Collect all non-trivial constant operations.
+  for (auto *N : nodes) {
+    // Skip trivial nodes/operations that do not require any constant
+    // computations.
+    if (llvm::isa<Storage>(N) || llvm::isa<Constant>(N) ||
+        llvm::isa<SplatNode>(N)) {
+      continue;
+    }
+    // Skip nodes that are not constant operations.
+    if (!isConstantOperation(N, *backend)) {
+      continue;
+    }
+    // Add only a constant operation node whose value is used by at least
+    // one non constant-operation node, because no other bigger constant
+    // operation containing the current node can completely replace the result
+    // of its computation. Doing this check allows for performing a smaller
+    // number of evaluateConstantOperation calls later and thus reduces the
+    // overhead.
+    if (hasNonConsantOperationUser(N, *backend)) {
+      worklist.push_back(N);
+    }
+  }
+  // Compute the result of each collected non-trivial constant operation.
+  while (!worklist.empty()) {
+    auto *C = worklist.pop_back_val();
+    std::vector<Constant *> constResults =
+        evaluateConstantOperation(*backend, cctx, C);
+    // Replace all results of the original operation by the computed
+    // compile-time results of this operation.
+    for (unsigned idx = 0, e = constResults.size(); idx < e; ++idx) {
+      auto constResult = constResults[idx];
+      // Replace the old result by the new constant result.
+      C->getNthResult(idx).replaceAllUsesOfWith(constResult);
+      // Add all users of constResult to the worklist if they became constant
+      // operations.
+      for (auto &use : constResult->getUsers()) {
+        auto *user = use.getUser();
+        if (isConstantOperation(user, *backend)) {
+          worklist.push_back(user);
```

Comment:
I'm not sure we need the worklist pattern at this point actually given the updated use of `hasNonConstantOperationUser()`. Is it possible for there to be ops that are newly constant-foldable after we do a round of constant folding of what was initially put in the worklist?

Reply:
Yeah, I guess not. We can probably process the found constant nodes as we go without any worklist. The only scenario where it may fail is when new nodes are added to the function during constant folding (e.g. during lowering or something like that). But it probably highly unlikely.

Reply:
Removed the worklist.

Reply:
You still have some of the logic here left, i.e. iterating over users and pushing to the worklist. I think you can remove all of it at this point.